### PR TITLE
use 0.5 for missing codons

### PR DESCRIPTION
**Pull request by mdehoon**
In `CodonAdaptationIndex`, use a count of 0.5 for codons that do not appear in the reference sequences. See #243.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
Acknowledge @tsp-kucbd for the suggested fix?

**Comment by peterjc** [comment]
Does the test failure help?

```
  TAC	0.434
- TAG	0.062
?    	   ^^
+ TAG	0.000
?    	   ^^
  TAT	1.000
  TCA	1.000
```

**Comment by mdehoon** [comment]
Anybody know why on line 157 we are dividing by `(cai_length - 1.0)` instead of by `cai_length` ?

**Comment by mdehoon** [comment]
> Acknowledge @tsp-kucbd for the suggested fix?

Sure. You mean in the copyright statement, or in the `NEWS.rst` file?

**Comment by mdehoon** [comment]
> Does the test failure help?

The problem is that by fixing this bug, the calculated CAI may change, as in this test case. Rather than fixing this bug and surprising users with CAI values changing suddenly, I suggest that we keep the current code as is, deprecate it, and copy a modernized version of the code to `Bio/SeqUtils/__init__.py` . This will allow us to subclass `CodonAdaptationIndex` from `dict`, and to avoid needing a Fasta file for initialization. The deprecation warning can alert the users that the new code may produce slightly different (and now correct) results.

We still need to figure out if the denominator on line 157 is correct. If not, we can also fix it in the new code.

**Comment by mdehoon** [comment]
> We still need to figure out if the denominator on line 157 is correct. If not, we can also fix it in the new code.

Comparing the calculated CAI with the values in Sharp & Li Table 1, it seems indeed that the denominator on line 157 was incorrect. I have fixed it in the new code.

**Comment by mdehoon** [comment]
@peterjc  OK to proceed?

File: Bio/SeqUtils/__init__.py [hunk 1]
```diff
@@ -577,6 +578,131 @@ def six_frame_translations(seq, genetic_code=1):
...
+
+    Implements the codon adaptation index (CAI) described by Sharp and
+    Li (Nucleic Acids Res. 1987 Feb 11;15(3):1281-95).
+
+    NOTE - This implementation does not currently cope with alternative genetic
+    codes: only the synonymous codons in the standard table are considered.
+    """
+
+    # each tuple consists of codons encoding the same amino acid
+    synonymous_codons = (
```
Comment (peterjc) [inline Bio/SeqUtils/__init__.py:592]: Feel like we could pull this from ``Bio/Data/CodonTable.py`` but not essential.

File: Bio/SeqUtils/__init__.py [hunk 2]
```diff
@@ -577,6 +578,131 @@ def six_frame_translations(seq, genetic_code=1):
...
 
 
+class CodonAdaptationIndex(dict):
+    """A codon adaptation index (CAI) implementation.
+
+    Implements the codon adaptation index (CAI) described by Sharp and
+    Li (Nucleic Acids Res. 1987 Feb 11;15(3):1281-95).
+
+    NOTE - This implementation does not currently cope with alternative genetic
+    codes: only the synonymous codons in the standard table are considered.
```
Comment (peterjc) [inline Bio/SeqUtils/__init__.py:588]: Perhaps in future ``__init__`` could take a ``table=...`` argument that could setup ``synonymous_codons`` at initialisation?

**Comment by peterjc** [comment]
I was thinking acknowledge Thomas in the NEWS file, but it looks like his one line fix has been over taken by a bigger refactoring.

File: Bio/SeqUtils/__init__.py (hunk 1 above)
Comment (mdehoon) [reply Bio/SeqUtils/__init__.py]: done

File: Bio/SeqUtils/__init__.py (hunk 2 above)
Comment (mdehoon) [reply Bio/SeqUtils/__init__.py]: done

**Review by peterjc — APPROVED** [review]

**Pull request merged**